src/EditorFeatures/Core/Implementation/ReferenceHighlighting/ReferenceHighlightingViewTaggerProvider.cs

```diff
@@ -2,87 +2,196 @@
+                TaggerEventSources.OnCaretPositionChanged(textView, textView.TextBuffer, TaggerDelay.Short),
+                TaggerEventSources.OnSemanticChanged(subjectBuffer, TaggerDelay.OnIdle, _semanticChangeNotificationService),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.Short),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.ReferenceHighlighting, TaggerDelay.NearImmediate));
+        }
 
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+        public override SnapshotPoint? GetCaretPoint(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return textViewOpt.Caret.Position.Point.GetPoint(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType), PositionAffinity.Successor);
+        }
 
-        protected override IEnumerable<PerLanguageOption<bool>> TagSourcePerLanguageOptions
+        public override IEnumerable<SnapshotSpan> GetSpansToTag(ITextView textViewOpt, ITextBuffer subjectBuffer)
         {
-            get
-            {
-                yield return FeatureOnOffOptions.ReferenceHighlighting;
-            }
+            return textViewOpt.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType))
+                              .Select(b => b.CurrentSnapshot.GetFullSpan())
+                              .ToList();
         }
 
-        protected override TimeSpan UIUpdateDelay
+        public override Task ProduceTagsAsync(TaggerContext<NavigableHighlightTag> context)
         {
-            get
+            // NOTE(cyrusn): Normally we'd limit ourselves to producing tags in the span we were
+            // asked about.  However, we want to produce all tags here so that the user can actually
+            // navigate between all of them using the appropriate tag navigation commands.  If we
+            // don't generate all the tags then the user will cycle through an incorrect subset.
+            if (context.CaretPosition == null)
             {
-                return TimeSpan.FromMilliseconds(TaggerConstants.NearImmediateDelay);
+                return SpecializedTasks.EmptyTask;
             }
+
+            var caretPosition = context.CaretPosition.Value;
+
+            Workspace workspace;
+            if (!Workspace.TryGetWorkspace(caretPosition.Snapshot.AsText().Container, out workspace))
+            {
+                return SpecializedTasks.EmptyTask;
+            }
+
+            var document = context.SpansToTag.First(vt => vt.SnapshotSpan.Snapshot == caretPosition.Snapshot).Document;
+            if (document == null)
+            {
+                return SpecializedTasks.EmptyTask;
+            }
+
+            // Don't produce tags if the feature is not enabled.
+            if (!workspace.Options.GetOption(FeatureOnOffOptions.ReferenceHighlighting, document.Project.Language))
```

Comment:
Do this as soon as we've got the workspace? Or is this dead now that taggers specify the options they trigger directly?